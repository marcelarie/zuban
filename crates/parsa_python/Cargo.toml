[package]
name = "parsa_python"
version.workspace = true
edition.workspace = true
rust-version.workspace = true
publish = false
homepage.workspace = true
authors.workspace = true

[dependencies]
parsa.workspace = true
lazy_static.workspace = true
regex.workspace = true

[dev-dependencies]
insta.workspace = true
test_utils.workspace = true
