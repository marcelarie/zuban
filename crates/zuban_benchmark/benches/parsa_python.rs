use criterion::{criterion_group, criterion_main, Criterion};

use parsa_python_cst::Tree;

fn parse_file(copies: usize) {
    let some_code = utils::dedent(
        r#"
        import os
        def x(a, b, c):
            try:
                return a
            except Exception:
                return [1, 2, 3, 4, 5, 6]
        class C:
            x: int
            y = 1 # type: int
            z: dict[int, list[str]] = {1: ["", r''], 2: [f""" {1} """]}
            def f(self, x: list[str], y: int): ...
    "#,
    );
    assert_eq!(some_code.len(), 262);
    Tree::parse(some_code.repeat(copies).into());
}

fn bench_parser(c: &mut Criterion) {
    c.bench_function("parse big file", |b| b.iter(|| parse_file(1000)));
    c.bench_function("parse small file", |b| b.iter(|| parse_file(1)));
}

// Register the benchmarks
criterion_group!(benches, bench_parser);
criterion_main!(benches);
