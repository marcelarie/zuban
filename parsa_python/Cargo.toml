[package]
name = "parsa_python"
version = "0.1.0"
edition.workspace = true
rust-version.workspace = true
homepage.workspace = true
authors.workspace = true

[dependencies]
parsa.workspace = true
lazy_static.workspace = true
regex.workspace = true

[dev-dependencies]
insta.workspace = true
