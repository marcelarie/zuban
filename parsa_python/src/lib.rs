use parsa::{create_parser,create_node,create_token};


create_token!(PythonToken, enum TokenType, String, Number, Endmarker);
create_node!(PythonNode, enum NodeType, file_input, foo);

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn it_works() {
        assert_eq!(2 + 2, 4);
        parse_python("bar");
        dbg!(TokenType::get_map());
        dbg!(NodeType::get_map());
        return
    }
}

struct PythonTokenizer {
}

impl PythonTokenizer {
    fn new_tok(&self) -> PythonToken {
        PythonToken {
            start: 0,
            length: 0,
            type_: TokenType::Endmarker,
        }
    }
}

impl parsa::Tokenizer<PythonToken> for PythonTokenizer {
    fn new(_code: &str) -> Self {
        Self {}
    }

    fn yield_next(&self) -> PythonToken {
        self.new_tok()
    }

}

create_parser!(parse_python, PythonTree, PythonNode,
               PythonToken, PythonTokenizer, TokenType, NodeType);
//create_parser!(parse_pythonx, PythonTreex, PythonNodex,
//               PythonToken, PythonTokenizer, TokenType, NodeType);
