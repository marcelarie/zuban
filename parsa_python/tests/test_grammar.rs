extern crate insta;

use parsa_python::*;

fn dedent(s: &'static str) -> String {
    let mut indent = usize::MAX;
    let lines: &Vec<_> = &s.split('\n').collect();
    for line in lines {
        if !line.trim().is_empty() {
            indent = std::cmp::min(indent, line.len() - line.trim_start().len());
        }
    }
    if indent == usize::MAX {
        return s.to_string();
    }
    let new_lines: Vec<_> = lines
        .iter()
        .map(|line| {
            if line.len() >= indent {
                &line[indent..]
            } else {
                line
            }
        })
        .collect();
    new_lines.join("\n")
}

fn tree_to_string(tree: PyTree) -> String {
    fn recurse(code: &mut String, node: &PyNode, depth: usize) {
        *code += &" ".repeat(depth);
        *code += &format!(
            "{}: {}-{}{}\n",
            node.type_str(),
            node.start(),
            node.end(),
            if node.is_leaf() {
                format!(" {:?}", node.as_code())
            } else {
                "".to_string()
            }
        );
        for c in node.iter_children() {
            assert_eq!(node.index, c.parent().unwrap().index);
            recurse(code, &c, depth + 1);
        }
    }

    let root = tree.root_node();
    assert!(root.parent().is_none());
    let mut code = String::new();
    recurse(&mut code, &root, 0);
    code
}

macro_rules! parametrize_snapshots {
    ($($name:ident : $input:expr;)*) => {$(
        #[test]
        fn $name() {
            let tree = parse($input.into());
            insta::assert_snapshot!(stringify!($name), tree_to_string(tree));
        }
    )*}
}

parametrize_snapshots!(
    empty: "".to_owned();
    very_simple: "1".to_owned();
    json: "{foo: 1}\n".to_owned();
    simple: dedent("
        a(bar:=1, foo=1);
        if a:
         1
        else: 2
        ");
    assignments_simple: dedent("
        a = 1
        a, b = 2
        a, b, = 2
        (a, b) = 2
        ((a, b)) = 2
        a = b = 3
        a = (b, c) = 3
        a = (b, c,) = 3
        a = b, c, = 3
        ");
    assignments_star: dedent("
        *foo = *bar = *baz
        i, *foo = ((j, [*bar, k])) = *baz
        *foo, *bar = yield baz
        ");
    assignments_aug: dedent("
        foo += 3
        foo.bar |= 3
        foo, bar += 3
        ");
    assignments_setitem: dedent("
        foo[0] += 3
        foo[0] = a = 3
        foo[a, b] = c
        foo[a::, b:f, c:=3] = c
        ");
    assignments_annotation: dedent("
        foo: bar
        foo: int = 3
        foo.bar: baz
        foo: bar = a = 4
        ");
    assignments_with_call: dedent("
        foo().bar = 1
        foo(1,).bar[1] = 1
        foo(a for a in b).bar[1] = 1
        ");
    for_stmt: dedent("
        for x in [1,2]:
            ...
        for *b in *c: ...
        for c.x, (*b, d.y[0]) in i, *j, k: ...
        ");
    comprehension_simple: dedent("
        [1 for a in b]
        [a for *a, c in b if c if d]
        foo(a for (*a, (*c, b),), in d)
        ");
    comprehension_fails: dedent("
        # fails
        [1 for a in *b]
        [1 for a in b if a else c]
        ");
    del_stmt: dedent("
        del foo
        del foo,
        del foo, bar.baz
        del foo[1]
        ");
    cls: dedent("
        class Foo(object):
            def __init__(self, /, f, *, g):
                pass

            @property
            def prop(self):
                pass
        ");
    import_names: dedent("
        import foo
        import foo.bar as baz
        import foo, bar, baz
        import foo.bar.baz
        ");
    import_name_fails: dedent("
        import foo,
        import foo, bar,
        ");
    import_from_normal: dedent("
        from foo import bar
        from foo import bar as baz
        from foo.bar.baz import *
        ");
    import_from_dotted: dedent("
        from . import (foo as baz, bar)
        from . import (foo, bar,)
        from ..foo import bar as baz
        from ... import *
        ");
    import_from_fails: dedent("
        from . import foo,
        from . import foo, bar,
        ");
    calls: dedent("
        foo(bar)
        foo(bar_foo=3)
        foo(bar_baz:=2)
        foo(**kwargs)
        ");
    nested_call: dedent("
        foo(1)()
        ");
    complex_calls: dedent("
        foo(bar, 3)
        foo(bar, *foo)
        foo(bar, *foo, name=3)
        foo(bar, *foo, name=3, **kwargs)
        foo(bar, *foo, name=3, **kwargs, b=3)
        foo(**kwargs, b=3, **foo)
        ");
    operators: dedent("
        f | d | c + 5 & 2 * 3 * 4 ** 5 ** 6
        ");
    simple_error_recovery: dedent("
        a + 3 +
        b = 3
        ");
    terminal_error_recovery: dedent("
        ?
        ");
    terminal_and_nonterminal_error_recovery: dedent("
        foo
        (1) + ? + c
        bar
        ");
    keyword_recovery: dedent("
        foo
        def
        bar
        else:
        baz
        ");
    match_simple: dedent("
        match foo:
            case ['a']: ...
            case _:
                pass
        ");
    soft_keyword_underscore: dedent("
        match _:
            case ['a']: _(_)
            case _:
                _
        match = 3
        ");
    match_underscore_with_as: dedent("
        match foo:
            case bar as _: pass
        ");
    match_class_pattern: dedent("
        match a:
            case A():
                pass
            case B(1, 2):
                pass
            case B(1, b=2):
                pass
            case B(a=1, b=2):
                pass
        ");
    dict_literal1: dedent("
        {1: 2}
        {**foo}
        {**foo, 1: 2}
        {**foo, **bar}
        ");
    function_simple: dedent("
        def foo(): ...
        def foo(a: 3): ...
        def foo(a, b,): ...
        def foo(a: f = 3, b=4): ...
        def foo(a=3, b=4): ...
        ");
    function_args: dedent("
        def foo(a, *, b=3): ...
        def foo(a=3, *, c): ...
        def foo(a, *, x, **kwargs): ...
        def foo(a, *args: int, **fooo: str,): ...
        def foo(a, /, b: i, c: i=3, **kwargs): ...
        def foo(*args): ...
        ");
    lambda_simple: dedent("
        lambda: ...
        lambda a: ...
        lambda a, b,: ...
        lambda a=3, b=4: ...
        ");
    lambda_slash: dedent("
        lambda a, /, b=3: ...
        lambda a=3, /,: ...
        lambda a=3, /: ...
        lambda a=3, /, *args, c=3: ...
        lambda a=3, /, c=3: ...
        lambda a, /, b, c=3, **kwargs: ...
        lambda a, a=3, /, c=3, *, d: ...
        ");
    lambda_args: dedent("
        lambda a, *, b=3: ...
        lambda a=3, *, c: ...
        lambda a, *, x, **kwargs: ...
        lambda a, *args, **fooo ,: ...
        lambda *args: ...
        ");
    lambda_kwargs: dedent("
        lambda **kwargs: ...
        lambda a=3, **kwargs: ...
        lambda a, **kwargs: ...
        lambda a, /, b, *, x, **kwargs: ...
        ");
    lambda_fails: dedent("
        lambda a, *, **kwargs: ...
        lambda a=3, b: ...
        lambda **a, b: ...
        lambda a=3, *,: ...
        lambda ,: ...
        ");
    lambda_fail_with_slash: dedent("
        lambda a=3, /, b: ...
        lambda a=3, /, b=3, c: ...
        lambda a, /, b=3, c: ...
        lambda z, a=3, d, /, b: ...
        ");
    with_items_simple: dedent("
        with foo as bar: ...
        with foo as bar, baz: ...
        ");
    with_items_parentheses: dedent("
        with (foo as bar): ...
        with (foo as bar, baz): ...
        with ((foo)): ...
        with (i for i in 42): ...
        with (yield y): ...
        ");
    string_literals_in_strings: dedent(r#"
        '"""'
        "'''"
        ""
        ''
        "#);
    bytes: dedent(r#"
        b'asdf\n\''
        br'\asdf'
        Br'asdf"""'
        bR''
        BR'\''
        "#);
    exception_group: dedent(r#"
        try:
            ...
        except* SpamError:
            ...
        except* FooError as e:
            ...
        except* (BarError, BazError) as e:
            ...
        "#);
    invalid_syntax: dedent(r#"
        def f():
            (x,) += 1
            z
        "#);
);
