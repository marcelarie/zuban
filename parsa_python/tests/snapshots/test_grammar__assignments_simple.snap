---
source: parsa_python/tests/test_grammar.rs
expression: tree_to_string(tree)
---
file: 0-60
 stmt: 1-7
  simple_stmts: 1-7
   simple_stmt: 1-6
    assignment: 1-6
     star_targets: 1-2
      star_target: 1-2
       target_with_star_atom: 1-2
        star_atom: 1-2
         name_definition: 1-2
          Name: 1-2 "a"
     Keyword: 3-4 "="
     star_expressions: 5-6
      expression: 5-6
       bitwise_or: 5-6
        Number: 5-6 "1"
   Newline: 6-7 "\n"
 stmt: 7-16
  simple_stmts: 7-16
   simple_stmt: 7-15
    assignment: 7-15
     star_targets: 7-11
      star_target: 7-8
       target_with_star_atom: 7-8
        star_atom: 7-8
         name_definition: 7-8
          Name: 7-8 "a"
      Keyword: 8-9 ","
      star_target: 10-11
       target_with_star_atom: 10-11
        star_atom: 10-11
         name_definition: 10-11
          Name: 10-11 "b"
     Keyword: 12-13 "="
     star_expressions: 14-15
      expression: 14-15
       bitwise_or: 14-15
        Number: 14-15 "2"
   Newline: 15-16 "\n"
 stmt: 16-26
  simple_stmts: 16-26
   simple_stmt: 16-25
    assignment: 16-25
     star_targets: 16-21
      star_target: 16-17
       target_with_star_atom: 16-17
        star_atom: 16-17
         name_definition: 16-17
          Name: 16-17 "a"
      Keyword: 17-18 ","
      star_target: 19-20
       target_with_star_atom: 19-20
        star_atom: 19-20
         name_definition: 19-20
          Name: 19-20 "b"
      Keyword: 20-21 ","
     Keyword: 22-23 "="
     star_expressions: 24-25
      expression: 24-25
       bitwise_or: 24-25
        Number: 24-25 "2"
   Newline: 25-26 "\n"
 stmt: 26-37
  simple_stmts: 26-37
   simple_stmt: 26-36
    assignment: 26-36
     star_targets: 26-32
      star_target: 26-32
       target_with_star_atom: 26-32
        star_atom: 26-32
         Keyword: 26-27 "("
         star_targets: 27-31
          star_target: 27-28
           target_with_star_atom: 27-28
            star_atom: 27-28
             name_definition: 27-28
              Name: 27-28 "a"
          Keyword: 28-29 ","
          star_target: 30-31
           target_with_star_atom: 30-31
            star_atom: 30-31
             name_definition: 30-31
              Name: 30-31 "b"
         Keyword: 31-32 ")"
     Keyword: 33-34 "="
     star_expressions: 35-36
      expression: 35-36
       bitwise_or: 35-36
        Number: 35-36 "2"
   Newline: 36-37 "\n"
 stmt: 37-50
  simple_stmts: 37-50
   simple_stmt: 37-49
    assignment: 37-49
     star_targets: 37-45
      star_target: 37-45
       target_with_star_atom: 37-45
        star_atom: 37-45
         Keyword: 37-38 "("
         star_targets: 38-44
          star_target: 38-44
           target_with_star_atom: 38-44
            star_atom: 38-44
             Keyword: 38-39 "("
             star_targets: 39-43
              star_target: 39-40
               target_with_star_atom: 39-40
                star_atom: 39-40
                 name_definition: 39-40
                  Name: 39-40 "a"
              Keyword: 40-41 ","
              star_target: 42-43
               target_with_star_atom: 42-43
                star_atom: 42-43
                 name_definition: 42-43
                  Name: 42-43 "b"
             Keyword: 43-44 ")"
         Keyword: 44-45 ")"
     Keyword: 46-47 "="
     star_expressions: 48-49
      expression: 48-49
       bitwise_or: 48-49
        Number: 48-49 "2"
   Newline: 49-50 "\n"
 stmt: 50-60
  simple_stmts: 50-60
   simple_stmt: 50-59
    assignment: 50-59
     star_targets: 50-51
      star_target: 50-51
       target_with_star_atom: 50-51
        star_atom: 50-51
         name_definition: 50-51
          Name: 50-51 "a"
     Keyword: 52-53 "="
     star_targets: 54-55
      star_target: 54-55
       target_with_star_atom: 54-55
        star_atom: 54-55
         name_definition: 54-55
          Name: 54-55 "b"
     Keyword: 56-57 "="
     star_expressions: 58-59
      expression: 58-59
       bitwise_or: 58-59
        Number: 58-59 "3"
   Newline: 59-60 "\n"
 Endmarker: 60-60 ""

